### DISCUSSION TITLE
Question: how does Pluto run notebooks with different Pkg Environments?

### DISCUSSION THREAD

**Discussion by zahachtah** [fonsp/Pluto.jl#1712] [Pluto development] [answered]
I am curious how Pluto accomplishes this. As I understand a notebook runs in a process with id=pid from a worker pool using the distributed package. 

I saw this comment:

" # if multiple notebooks run on the same process, then we need to `cd` between the different notebook paths"

and then using the function:

use_nbpkg_environment

to switch paths to the environment(?)

Do I understand this correctly if this means that if two notebooks use the same pid and both are actively evaluating cells at the same time the process has to switch pkg environment very frequently?

Mostly curious about how Pluto does things so smartly :-)

**Comment by Pangoraw** [comment, marked as answer]
There is only one notebook per process. The comment is refering to the case where you launch pluto with `workspace_use_distributed=false` (`workspace.pid == Distributed.myid()` means that the notebook is living in the same process as Pluto itself) which is mainly used for faster testing time (see https://github.com/fonsp/Pluto.jl/discussions/1841#discussioncomment-2057440).
